### ISSUE TITLE
Updated from to 5.0.6 from 4.4.2

### ISSUE THREAD

**Issue by mmdock** [SDWebImage/SDWebImage#2771]
### New Issue Checklist

* [X] I have read and understood the [CONTRIBUTING guide](https://github.com/rs/SDWebImage/blob/master/.github/CONTRIBUTING.md)
* [X] I have read the [Documentation](http://cocoadocs.org/docsets/SDWebImage/)
* [X] I have searched for a similar issue in the [project](https://github.com/rs/SDWebImage/issues) and found none

### Issue Info

 Info                    | Value                               |
-------------------------|-------------------------------------|
 Platform Name           | ios 
 Platform Version        | 12.3.1
 SDWebImage Version      | 5.0.6
 Integration Method      | cocoapods 1.5.3
 Xcode Version           | Version 10.2.1 (10E1001)
 Repro rate              | 100%

### Issue Description and Steps
I had the following class being used in my project with SDWebImage 4.4.2:

```
class ImageView: UIImageView {

    private var operation:SDWebImageOperation?
    private let avatar: Bool

    private let cornerRadius: CGFloat?
    
    override var image: UIImage? {
        didSet {
            overlayView?.isHidden = (image == nil || overlayHidden)
            loader.isHidden = image != nil
            if let overlayView = overlayView {
                bringSubviewToFront(overlayView)
            }
        }
    }
    var overlayHidden:Bool = false {
        didSet {
            overlayView?.isHidden = overlayHidden
        }
    }
    let loader = ImageLoaderView()
    private let overlayView: OverlayView?

    init(type: OverlayType = .none, avatar: Bool = false, _ cornerRadius:CGFloat = 25) {
        self.avatar = avatar
        self.cornerRadius = cornerRadius
        overlayView = OverlayView(type: type)
        super.init(frame: CGRect())
        self.customInit()
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        if avatar {
            layer.cornerRadius = self.bounds.size.width / 2
        }
    }
    
    func customInit(){
        backgroundColor = UIColor.herLightGray
        contentMode = .scaleAspectFill
        clipsToBounds = true
        
        if avatar {
            layer.cornerRadius = self.cornerRadius ?? 25
        }
        addSubview(loader)
        overlayView?.isHidden = true
        if let overlayView = overlayView {
            addSubview(overlayView)
        }
        loader.snp.makeConstraints { make in
            make.center.equalTo(self)
        }
        overlayView?.snp.makeConstraints { make in
            make.edges.equalTo(self)
        }
    }

    required init?(coder aDecoder: NSCoder) {
        avatar = false
        cornerRadius = 25
        overlayView = OverlayView(type: .none)
        super.init(coder: aDecoder)
        self.customInit()
    }

    func setImage(_ url: URL?, withDeviceSize: Bool = false, asTemplate: Bool = false, completed: SDExternalCompletionBlock? = nil) {
        loader.progress = 0

        var urlToLoad = url
        if withDeviceSize{
            urlToLoad = url?.withSize(.device)
        }else{
            urlToLoad = url?.withSize(avatar ? .small : .large)
        }

        operation = SDWebImageManager.shared.loadImage(with: urlToLoad , options: .highPriority, progress: {[weak self]  (received, expected, _) in
            if expected > 0 && received > 0 && received <= expected {
                DispatchQueue.main.async {
                    self?.loader.progress = Double(received) / Double(expected)
                }
            }
        }) {[weak self] (image, data, error, cacheType, _, imageURL) in
            DispatchQueue.main.async {
                self?.loader.isHidden = true
                if !asTemplate{
                    self?.image = image
                }else{
                    self?.image = image?.withRenderingMode(.alwaysTemplate)
                }
            }
            completed?(image, error, cacheType, imageURL)
        }
    }
    
    func cancelLoading(){
        operation?.cancel()
        image = nil
    }

```

Upon updating to 5.0.6 this class no longer works to download any images.  

I was reading through the migration guide and did not notice anything that could affect this: https://github.com/SDWebImage/SDWebImage/wiki/5.0-Migration-guide

Could use some assistance on how to get this back up and running.

**Comment by dreampiggy** [comment]
By Using your code with a little demo, I can successfully load the image and display.

See:
![image](https://user-images.githubusercontent.com/6919743/60150463-723a5b00-980b-11e9-97c7-d58a5d6b00db.png)

Some of category is not available, but I don't think they'll effect the result.

So, what you can do is check the whether the final URL is correct (for example,  the `URL.withSize(:)` logic. Other View code seems does not effect the loading result.

**Comment by mmdock** [comment]
@dreampiggy 

It seems as if the below block of code is not firing requests.  I have enabled Charles proxy on my phone to see all outgoing requests and with the update to 5.0.6 no requests are getting made by this block.

The completion block is getting called with an image of 0 pixels and no data.  However, i never log any outbound requests (which do show up with the working 4.4.2 implementation).

update for SDWebImage from 4.4.2 to 5.0.6 that i had done:

Transitive requirement of FLAnimatedImage was removed for SDWebImageFLPlugin

SDWebImage/WebP became SDWebImageWebPCoder

libwebp updated from 0.6.1 to 1.0.2

all instances of `shared()` became `shared`

`maxCacheAge` became `maxDiskAge`

App Boy updated to 3.15.0 from 3.9.0.

Am i missing some initial configuration that SDWebImage 5.0.X needs in order to perform correctly?

```
        operation = SDWebImageManager.shared.loadImage(with: urlToLoad , options: .highPriority, progress: {[weak self]  (received, expected, _) in
            if expected > 0 && received > 0 && received <= expected {
                DispatchQueue.main.async {
                    self?.loader.progress = Double(received) / Double(expected)
                }
            }
        }) {[weak self] (image, data, error, cacheType, _, imageURL) in
            DispatchQueue.main.async {
                self?.loader.isHidden = true
                if !asTemplate{
                    self?.image = image
                }else{
                    self?.image = image?.withRenderingMode(.alwaysTemplate)
                }
            }
            completed?(image, error, cacheType, imageURL)
        }
```

**Comment by dreampiggy** [comment]
> The completion block is getting called with an image of 0 pixels and no data

Provide your image url, and set a breakpoint of the decoding part (Just set it in `SDWebImageCodersManager.decodedImageForData:options:`).

If you think it's something not related to Network, but only about the image decoding, we can simplify the problem to only focus on this part. Don't need to make things complicated. (Such as View class, cache, etc)

**Comment by dreampiggy** [comment]
And, if you use some specify image format (WebP, etc). Don't forget to register the image coder in your App launch time.

```
// AppDelegate.swift
func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey : Any]? = nil) -> Bool
    SDImageCodersManager.shared.add(coder: SDImageWebPCoder.shared)
}
```

**Comment by mmdock** [comment]
@dreampiggy thanks!  i just needed SDImageCodersManager.shared.add(coder: SDImageWebPCoder.shared)

Strange to me that Charles Proxy still isn't logging the image download requests anymore.  But it is certainly downloading the images now.

**Comment by mmdock** [comment]
Hey @dreampiggy 

I recently released this change as all image downloading seemed to be working correctly (for images and GIFs).

The issue I am now having though is: 
Fatal Exception: NSInvalidArgumentException
0  CoreFoundation                 0x1d689127c __exceptionPreprocess
1  libobjc.A.dylib                0x1d5a6b9f8 objc_exception_throw
2  CoreFoundation                 0x1d679b4b0 -[NSCache init]
3  Foundation                     0x1d72fc468 NSAllocateMemoryPages
4  Foundation                     0x1d72b13e0 _NSDataCreateVMDispatchData
5  Foundation                     0x1d71f0c6c -[_NSPlaceholderData initWithBytes:length:copy:deallocator:]
6  SDWebImage                     0x105a14384 -[SDWebImageDownloaderOperation URLSession:task:didCompleteWithError:] + 410 (SDWebImageDownloaderOperation.m:410)
7  SDWebImage                     0x105a10f98 -[SDWebImageDownloader URLSession:task:didCompleteWithError:] + 443 (SDWebImageDownloader.m:443)
8  CFNetwork                      0x1d7014420 __51-[NSURLSession delegate_task:didCompleteWithError:]_block_invoke.182
9  Foundation                     0x1d72f5ef8 __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__
10 Foundation                     0x1d72023e0 -[NSBlockOperation main]
11 Foundation                     0x1d72018c8 -[__NSOperationInternal _start:]
12 Foundation                     0x1d72f7c7c __NSOQSchedule_f
13 libdispatch.dylib              0x1d62d0a38 _dispatch_call_block_and_release
14 libdispatch.dylib              0x1d62d17d4 _dispatch_client_callout
15 libdispatch.dylib              0x1d62a8c34 _dispatch_continuation_pop$VARIANT$armv81
16 libdispatch.dylib              0x1d62a8314 _dispatch_async_redirect_invoke
17 libdispatch.dylib              0x1d62b49d4 _dispatch_root_queue_drain
18 libdispatch.dylib              0x1d62b5248 _dispatch_worker_thread2
19 libsystem_pthread.dylib        0x1d64b11b4 _pthread_wqthread
20 libsystem_pthread.dylib        0x1d64b3cd4 start_wqthread

is now a very prolific crash in my app in version 5.0.6 when a very similar crash from 4.4.2 had almost no impact.

Before the the updated, the very similar crash was from version 4.4.2:
Fatal Exception: NSInvalidArgumentException
0  CoreFoundation                 0x1828b3ea0 __exceptionPreprocess
1  libobjc.A.dylib                0x181a85a40 objc_exception_throw
2  CoreFoundation                 0x1827ba674 -[NSCache init]
3  Foundation                     0x1833491b0 NSAllocateMemoryPages
4  Foundation                     0x1832fb624 _NSDataCreateVMDispatchData
5  Foundation                     0x183239188 -[_NSPlaceholderData initWithBytes:length:copy:deallocator:]
6  SDWebImage                     0x105d7e8a4 -[SDWebImageDownloaderOperation URLSession:task:didCompleteWithError:] + 407 (SDWebImageDownloaderOperation.m:407)
7  SDWebImage                     0x105d7c0a0 -[SDWebImageDownloader URLSession:task:didCompleteWithError:] + 381 (SDWebImageDownloader.m:381)
8  CFNetwork                      0x18304d798 __51-[NSURLSession delegate_task:didCompleteWithError:]_block_invoke.241
9  Foundation                     0x18334282c __NSBLOCKOPERATION_IS_CALLING_OUT_TO_A_BLOCK__
10 Foundation                     0x18324aa28 -[NSBlockOperation main]
11 Foundation                     0x183249efc -[__NSOperationInternal _start:]
12 Foundation                     0x183344700 __NSOQSchedule_f
13 libdispatch.dylib              0x1822ed6c8 _dispatch_call_block_and_release
14 libdispatch.dylib              0x1822ee484 _dispatch_client_callout
15 libdispatch.dylib              0x18229183c _dispatch_continuation_pop$VARIANT$mp
16 libdispatch.dylib              0x182290f04 _dispatch_async_redirect_invoke
17 libdispatch.dylib              0x18229da28 _dispatch_root_queue_drain
18 libdispatch.dylib              0x18229e2d0 _dispatch_worker_thread2
19 libsystem_pthread.dylib        0x1824d017c _pthread_wqthread
20 libsystem_pthread.dylib        0x1824d2cec start_wqthread

Here is the thing: 

in 4.4.2 this crash was happening about 20 times a day.

Now, I am seeing about 400 per day of that new one with the release that included 5.0.6.

The 4.4.2 crash, was primarily on my iOS 11 or iOS 10 users, hence I didn't care so much.   (something like 27 % of the crashes were for iOS 12)

But the 5.0.6 crash is 99% my iOS 12 users.

This, to me, is a very bizarre data point. 

TL;DR: 5.0.6 version appears to be causing higher volume of what appears to be OOM crashes on newer iOS OS compared to older 4.X.X version.

**Comment by dreampiggy** [comment]
Still OOM issue.

Did you use 4.x SDWebImage with some extra config, but you did not translate and migrate to new API ?
For example, some options may effect RAM usage:

+ shouldDecompressImages(force decode) -> Now is `SDWebImageAvoidDecodeImage` option per image-request
+ SDImaegCache.maxMemoryCost -> Now is `SDImageCahceConfig.maxMemoryCost` and the value should be like 4 times compared to original value (because we use used **Bytes** as memory cost, but not **Pixel Count**)
+

**Comment by mmdock** [comment]
@dreampiggy The ImageView class above is exactly what i used in 4.x (but it was previously `shared()` instead of `shared`)

List of exact changes i made:
Transitive requirement of FLAnimatedImage was removed for SDWebImageFLPlugin

SDWebImage/WebP became SDWebImageWebPCoder

libwebp updated from 0.6.1 to 1.0.2

all instances of `shared()` became `shared`

maxCacheAge became maxDiskAge and no change in value.

App Boy updated to 3.15.0 from 3.9.0. (it depended on SDWebImage 5.x.x)

annnndd per the original investigation on this thread i added: SDImageCodersManager.shared.addCoder(SDImageWebPCoder.shared)

that is the entire summary of my changes.

I am currently about to release a build that adds `.scaleDownLargeImages` flag as an option to see what that does.
